docs/naps/3-spaces.md

```diff
@@ -0,0 +1,127 @@
+
+# Implementation Proposal
+
+## API
+
+`ViewerModel.spaces` would be a (selectable) `EventedList` or similar evented collection, each containing a `Space` object, with the following attributes:
+- `name`: unique name for the space
+- `layers`: a `layerlist` (or top-level `layergroup`, in the future)
+- `active`: similar to how a layer in a `layerlist` can be active, a `Space` in the `spaces` can be active; this will determine which space is loaded into the layerlist and used to populate the canvas
+- `camera`: a snapshot of the `Camera` model
+- `dims`: a snapshot of the `Dims` model
+- ...
+
+In general, the most intuitive approach would be to mirror the basic `Layer` api (i.e: with `name`, `metadata`, possibly `source`...).
+
+All `spaces` would do is effectively provide a quick way to swap the `layerlist`, and update the `Camera` and the `Dims` models.
+
+## GUI
+
+The GUI could expose this as a (searchable) dropdown menu above the layerlist, and provide shortcuts to navigate easility through spaces, such as `page-up`, `page-down`.
+
+## Plugins
+
+Reader/writer plugins should be able to provide/consume spaces. If unspecified, they should act on the active `Space`, which would be backwards compatible.
+
+Widget plugins would be backwards compatible, as they simply act on the active `Space`. On the other hand, new plugins would be able to access other spaces as well, allowing for easier abstraction of "batch" workflows.
+
+
+# Alternatives
+
+## Multiple viewers
+
+These problems could be also solved by allowing multiple `Viewer` objects, each with its own `ViewerModel`, by separating out the `QtViewer` logic to an `Application` level [^application].
+
+```python
+app.viewers # list of all viewers
+app.viewer # current viewer model
+app.window # the singleton window now lives on the app rather than the viewer
+```
+
+This is in many ways equivalent to `spaces`, with object serving the same purpose but being named differently: `Application` takes place of the `Viewer` object; `Viewers` are acting as `spaces`, with the difference that they are themselves a `ViewerModel`, rather than holding a snapshot of parts of a `ViewerModel`.
+
+While the nomenclature is not one of the important points of this NAP, I find the use of the word `Viewer` confusing in this alternative implementation. Intuitively, for me (and, I expect, for most users) the `Viewer` is the window, regardless of the abstraction of `ViewerModel` and `Viewer` that we have on the backend. Additionally, the above changes would cause a big API change in the most basic interaction with `napari` (`app.viewer` rather than `viewer`), unless we "hide" it away (`viewer.app.viewers`), which would hurt usability and discoverability.
+
+Additionally, there are no benefits to keeping multiple `ViewerModel` objects alive, since their purpose is simply to act as an evented model to update GUI.
```

Comment:
```suggestion
Additionally, there are no benefits to keeping multiple `ViewerModel` objects alive, since their purpose is simply to act as an evented model to update the GUI.
```

Reply:
That's a good question, I didn't think about those... My gust reaction is *yes*, they should store that state; in fact, it might be as simple as doing (pseudocode):
```python
viewer.spaces[0] = viewer.dict()
...
viewer.update(viewer.spaces[0])
```

I'm also starting to be convinced by the "states" name argument, since it's clearly more than just a "physical" space that we need to track.

And yes, I realize that at this point we are simply reproducing the full `ViewerModel`, which kinda weakens my argument against using "multiple viewers"... I still think it's better to not keep many `ViewerModel` objects (the you have to deal with connecting/disconnecting events, the slowness of `EventedModel` instantiation, ...); also, I still find the naming scheme of that one just too confusing.

Reply:
> you have to deal with connecting/disconnecting events, the slowness of EventedModel instantiation

Does having multiple `Viewer` objects mean this? How is it different from a space (`~=ViewerModel`) which also needs to update the GUI?

Reply:
I added this to the API proposal, with some space left for discussion. I will update again in a follow up PR once people have given their opinion.